nvflare/fuel/f3/sfm/conn_manager.py

```diff
@@ -0,0 +1,452 @@
+        if not endpoint_name:
+            raise CommError(CommError.BAD_DATA, f"Handshake without endpoint name for connection {sfm_conn.get_name()}")
+
+        if endpoint_name == self.local_endpoint.name:
+            raise CommError(
+                CommError.BAD_DATA, f"Duplicate endpoint name {endpoint_name} for connection {sfm_conn.get_name()}"
+            )
+
+        endpoint = Endpoint(endpoint_name, data)
+        endpoint.state = EndpointState.READY
+        conn_props = sfm_conn.conn.get_conn_properties()
+        if conn_props:
+            endpoint.conn_props.update(conn_props)
+
+        sfm_endpoint = self.endpoints.get(endpoint_name)
+        if sfm_endpoint:
+            old_state = sfm_endpoint.endpoint.state
+            sfm_endpoint.endpoint = endpoint
+        else:
+            old_state = EndpointState.IDLE
+            sfm_endpoint = SfmEndpoint(endpoint)
+
+        sfm_endpoint.add_connection(sfm_conn)
+        sfm_conn.sfm_endpoint = sfm_endpoint
+        self.endpoints[endpoint_name] = sfm_endpoint
+
+        if endpoint.state != old_state:
+            self.notify_monitors(endpoint)
+
+    def notify_monitors(self, endpoint: Endpoint):
+
+        if not self.monitors:
+            log.debug("No endpoint monitor registered")
+            return
+
+        for monitor in self.monitors:
+            monitor.state_change(endpoint)
+
+    @staticmethod
+    def get_dict_payload(prefix, frame):
+        mv = memoryview(frame)
+        return msgpack.unpackb(mv[(PREFIX_LEN + prefix.header_len) :])
+
+    def handle_new_connection(self, connection: Connection):
+
+        sfm_conn = SfmConnection(connection, self.local_endpoint)
+        with self.lock:
+            self.connections[sfm_conn.get_name()] = sfm_conn
+
+        connection.register_frame_receiver(SfmFrameReceiver(self, sfm_conn))
+
+        if connection.connector.mode == Mode.ACTIVE:
+            sfm_conn.send_handshake(Types.HELLO)
+
+    def close_connection(self, connection: Connection):
+
+        with self.lock:
+            name = connection.name
+            if name not in self.connections:
+                log.debug(f"Connection {name} has closed with no endpoint assigned")
+                return
+
+            sfm_conn = self.connections.pop(name)
+            sfm_endpoint = sfm_conn.sfm_endpoint
+            if sfm_endpoint is None:
+                log.debug(f"Connection {name} is closed before SFM handshake")
+                return
+
+            old_state = sfm_endpoint.endpoint.state
+            sfm_endpoint.remove_connection(sfm_conn)
+
+            state = EndpointState.READY if sfm_endpoint.connections else EndpointState.DISCONNECTED
+            sfm_endpoint.endpoint.state = state
+            if old_state != state:
+                self.notify_monitors(sfm_endpoint.endpoint)
+
+    def send_loopback_message(self, endpoint: Endpoint, app_id: int, headers: Headers, payload: BytesAlike):
```

Comment:
Since we define the `Message` class, can we change the signature to use Message instead of headers and payload

Reply:
Message is an user-facing class, not used internally.